PyInstaller/building/api.py

```diff
@@ -254,8 +255,9 @@ def assemble(self):
                     seenFnms[fnm] = inm
                     seenFnms_typ[fnm] = typ
 
+                    do_upx = os.path.basename(fnm) not in self.exclude_upx
                     fnm = checkCache(fnm, strip=self.strip_binaries,
-                                     upx=(self.upx_binaries and (is_win or is_cygwin)),
+                                     upx=(self.upx_binaries and (is_win or is_cygwin) and do_upx),
```

Comment:
This becomes hard to read, given all these options. I suggest somthing like this:
```
do_upx = (self.upx_binaries
       and (is_win or is_cygwin)
       and os.path.basename(fnm) not in self.exclude_upx)
...
     up=do_upx,
```

`x not in list` is relatively expensive, thus this only needs to be checked if upx is to be used at all.